internal/memorylimiter/memorylimiter.go

```diff
@@ -0,0 +1,266 @@
+)
+
+// MemoryLimiter is used to prevent out of memory situations on the collector.
+type MemoryLimiter struct {
+	usageChecker memUsageChecker
+
+	memCheckWait time.Duration
+	ballastSize  uint64
+
+	// mustRefuse is used to indicate when data should be refused.
+	mustRefuse *atomic.Bool
+
+	ticker *time.Ticker
+
+	lastGCDone time.Time
+
+	// The function to read the mem values is set as a reference to help with
+	// testing different values.
+	readMemStatsFn func(m *runtime.MemStats)
+
+	// Fields used for logging.
+	logger                 *zap.Logger
+	configMismatchedLogged bool
+
+	refCounterLock sync.Mutex
+	refCounter     int
+	waitGroup      sync.WaitGroup
+	closed         chan struct{}
+}
+
+// NewMemoryLimiter returns a new memory limiter component
+func NewMemoryLimiter(cfg *Config, logger *zap.Logger) (*MemoryLimiter, error) {
+	usageChecker, err := getMemUsageChecker(cfg, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Info("Memory limiter configured",
+		zap.Uint64("limit_mib", usageChecker.memAllocLimit/mibBytes),
+		zap.Uint64("spike_limit_mib", usageChecker.memSpikeLimit/mibBytes),
+		zap.Duration("check_interval", cfg.CheckInterval))
+
+	return &MemoryLimiter{
+		usageChecker:   *usageChecker,
+		memCheckWait:   cfg.CheckInterval,
+		ticker:         time.NewTicker(cfg.CheckInterval),
+		readMemStatsFn: runtime.ReadMemStats,
+		logger:         logger,
+		mustRefuse:     &atomic.Bool{},
+	}, nil
+}
+
+// StartMonitoring starts a single ticker'd goroutine per instance
+// that will check memory usage every checkInterval period.
+func (ml *MemoryLimiter) StartMonitoring() {
+	ml.refCounterLock.Lock()
+	defer ml.refCounterLock.Unlock()
+
+	ml.refCounter++
+	if ml.refCounter == 1 {
+		ml.closed = make(chan struct{})
+		ml.waitGroup.Add(1)
+		go func() {
+			defer ml.waitGroup.Done()
+
+			for {
+				select {
+				case <-ml.ticker.C:
+				case <-ml.closed:
+					return
+				}
+				ml.checkMemLimits()
+			}
+		}()
+	}
+}
+
+// StopMonitoring resets MemoryLimiter monitoring ticker and stop monitoring
+func (ml *MemoryLimiter) StopMonitoring() error {
+	ml.refCounterLock.Lock()
+	defer ml.refCounterLock.Unlock()
+
+	if ml.refCounter == 0 {
+		return ErrShutdownNotStarted
+	} else if ml.refCounter == 1 {
+		ml.ticker.Stop()
+		close(ml.closed)
+		ml.waitGroup.Wait()
+	}
+	ml.refCounter--
+	return nil
+}
+
+// SetBallastSizeIfExtensionExists finds a memory ballast extension and get ballast size from the extension.
+// if no ballast extension is found, nothing is set
+func (ml *MemoryLimiter) SetBallastSizeIfExtensionExists(host component.Host) {
+	extensions := host.GetExtensions()
+	for _, extension := range extensions {
+		if ext, ok := extension.(interface{ GetBallastSize() uint64 }); ok {
+			ml.ballastSize = ext.GetBallastSize()
+			break
+		}
+	}
+}
+
+func (ml *MemoryLimiter) Start(_ context.Context, host component.Host) error {
+	ml.SetBallastSizeIfExtensionExists(host)
+	ml.StartMonitoring()
```

Comment:
These two functions don't need to be exported. Otherwise, it's a bit confusing.

Reply:
done